uniffi_bindgen/src/bindings/gecko/gen_gecko.rs

```diff
@@ -0,0 +1,355 @@
+// Note that this can only be used to control details *that do not affect the
+// underlying component*, since the details of the underlying component are
+// entirely determined by the `ComponentInterface`.
+pub struct Config {
+    // ...
+}
+
+impl Config {
+    pub fn from(_ci: &ComponentInterface) -> Self {
+        Config {
+            // ...
+        }
+    }
+}
+
+/// Indicates whether a WebIDL type is reflected as an out parameter or return
+/// value in C++. This is used by the namespace and interface templates to
+/// generate the correct argument lists for the binding.
+pub enum ReturnPosition<'a> {
+    OutParam(&'a Type),
+    Return(&'a Type),
+}
+
+impl<'a> ReturnPosition<'a> {
+    /// Given a type, returns a tag indicating whether it's returned by value or
+    /// via an out parameter.
+    pub fn for_type(type_: &'a Type) -> ReturnPosition<'a> {
+        match type_ {
+            Type::String => ReturnPosition::OutParam(type_),
+            Type::Optional(_) => ReturnPosition::OutParam(type_),
+            Type::Record(_) => ReturnPosition::OutParam(type_),
+            Type::Map(_) => ReturnPosition::OutParam(type_),
+            Type::Sequence(_) => ReturnPosition::OutParam(type_),
+            _ => ReturnPosition::Return(type_),
+        }
+    }
+}
+
+/// Returns a dummy empty value for the given type. This is used to
+/// implement the `cpp::bail` macro to return early if a WebIDL function or
+/// method throws an exception.
+pub fn default_return_value_cpp(type_: &Type) -> Option<String> {
+    Some(match type_ {
+        Type::Int8
+        | Type::UInt8
+        | Type::Int16
+        | Type::UInt16
+        | Type::Int32
+        | Type::UInt32
+        | Type::Int64
+        | Type::UInt64 => "0".into(),
+        Type::Float32 => "0.0f".into(),
+        Type::Float64 => "0.0".into(),
+        Type::Boolean => "false".into(),
+        Type::Enum(name) => format!("{}::EndGuard_", name),
+        Type::Object(_) => "nullptr".into(),
+        Type::String | Type::Record(_) | Type::Optional(_) | Type::Sequence(_) | Type::Map(_) => {
+            return None
+        }
+        Type::Error(name) => panic!("[TODO: ret_type_cpp({:?})]", type_),
+    })
+}
+
+/// A template for a Firefox WebIDL file. We only generate one of these per
+/// component.
+#[derive(Template)]
+#[template(syntax = "webidl", escape = "none", path = "WebIDLTemplate.webidl")]
+pub struct WebIdl<'config, 'ci> {
+    config: &'config Config,
+    ci: &'ci ComponentInterface,
+}
+
+impl<'config, 'ci> WebIdl<'config, 'ci> {
+    pub fn new(config: &'config Config, ci: &'ci ComponentInterface) -> Self {
+        Self { config: config, ci }
+    }
+}
+
+/// A shared header file that's included by all our bindings. This defines
+/// common serialization logic and `extern` declarations for the FFI. Note that
+/// the bindings always include this header file, never the other way around.
```

Comment:
...Because otherwise, we’d get a circular dependency when the bindings layer includes one of our interface headers in its header or implementation.